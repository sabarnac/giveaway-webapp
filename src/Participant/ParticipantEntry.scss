.participant-entry {
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  .participant-entry__avatar {
    text-align: center;
    height: 5rem;
    img {
      height: 5rem;
    }
    width: 50%;
  }
  .participant-entry__name {
    text-align: center;
    font-size: 3rem;
    width: 50%;
    word-break: break-all;
  }
  padding: 2rem;
  border: 1px solid #aaa;
  width: 45%;
}
