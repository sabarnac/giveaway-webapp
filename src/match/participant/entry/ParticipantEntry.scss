.participant-entry {
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    padding: 2rem;
    border: 1px solid #aaa;
    width: 100%;
    .participant-entry__avatar {
        text-align: center;
        height: 5rem;
        img {
            height: 5rem;
        }
    }
    .participant-entry__name {
        text-align: center;
        font-size: 3rem;
        flex-grow: 1;
        word-break: break-all;
    }
}