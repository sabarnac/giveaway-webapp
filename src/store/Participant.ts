import { observable, computed } from "mobx";

/**
 * Class containing the details of a participants' avatar.
 */
export class Avatar {
  @observable private _url: string;
  @observable private _altText: string;

  public constructor(url: string, altText: string) {
    this._url = url;
    this._altText = altText;
  }

  /**
   * Gets the avatar URL.
   * @return {string} The avatar URL.
   */
  public get url(): string {
    return this._url;
  }

  /**
   * Gets the avatar image alternate text.
   * @return {string} The avatar image alternate text.
   */
  public get altText(): string {
    return this._altText;
  }

  private isSameUrl = (url1: string, url2: string): boolean => url1 === url2;

  private isSameAltText = (altText1: string, altText2: string): boolean =>
    altText1 === altText2;

  /**
   * Determines whether another avatar is identical to the current one.
   * @param  {Avatar} otherAvatar The other avatar to compare against.
   * @returns {boolean} Whether the other avatar is equal to the current one.
   */
  public equals = (otherAvatar: Avatar): boolean =>
    this.isSameUrl(this._url, otherAvatar._url) &&
    this.isSameAltText(this._altText, otherAvatar._altText);
}

/**
 * Class containing the details of a participant.
 */
export default class Participant {
  @observable private _name: string;
  @observable private _avatar?: Avatar;

  public constructor(name: string, avatar?: Avatar) {
    this._name = name;
    this._avatar = avatar;
  }

  /**
   * Gets the participant name.
   * @return {string} The participant name.
   */
  public get name(): string {
    return this._name;
  }

  /**
   * Gets the participant avatar, if it exists.
   * @return {string | null} The participant avatar, or null if there isn't one.
   */
  public get avatar(): Avatar | null {
    return this._avatar || null;
  }

  /**
   * Gets whether the participant has an avatar or not.
   * @returns {boolean} Whether the participant has an avatar or not.
   */
  @computed public get hasAvatar(): boolean {
    return !!this._avatar;
  }

  /**
   * Gets the capitalized name of the participant.
   * @returns {string} The capitalized name.
   */
  @computed public get properName(): string {
    return this._name.replace(/\w\S*/g, (subText: string) => {
      return subText.charAt(0).toUpperCase() + subText.substr(1).toLowerCase();
    });
  }

  private isSameName = (name1: string, name2: string): boolean =>
    name1 === name2;

  private isSameAvatar = (avatar1?: Avatar, avatar2?: Avatar): boolean =>
    avatar1 && avatar2
      ? avatar1.equals(avatar2)
      : !avatar1 && !avatar2
      ? true
      : false;

  /**
   * Determines whether another participant is identical to the current one.
   * @param  {Participant} otherParticipant The other participant to compare against.
   * @returns {boolean} Whether the other participant is equal to the current one.
   */
  public equals = (otherParticipant: Participant): boolean =>
    this.isSameName(this._name, otherParticipant._name) &&
    this.isSameAvatar(this._avatar, otherParticipant._avatar);
}
