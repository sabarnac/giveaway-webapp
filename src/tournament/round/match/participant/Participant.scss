.participant {
    background: #ffffff;
    &--invert {
        background: #000000;
        color: #ffffff;
    }
}