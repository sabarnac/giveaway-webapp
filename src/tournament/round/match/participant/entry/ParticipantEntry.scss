@import '../Participant';
.participant-entry {
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    padding: 2rem;
    width: 100%;
    border: 1px solid #ffffff;
    border-bottom: none;
    &:last-child {
        border-bottom: 1px solid #ffffff;
    }
    &__avatar {
        text-align: center;
        height: 5rem;
        img {
            height: 5rem;
        }
    }
    &__name {
        text-align: center;
        font-size: 3rem;
        flex-grow: 1;
        word-break: break-all;
    }
}