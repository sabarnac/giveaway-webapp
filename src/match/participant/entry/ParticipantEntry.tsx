import React, { Component } from "react";
import { observer } from "mobx-react";
import "./ParticipantEntry.scss";
import classNames from "classnames";
import Participant from "../../../store/round/match/participant/Participant";

interface ParticipantEntryProps {
  participant: Participant;
}

@observer
class ParticipantEntry extends Component<ParticipantEntryProps> {
  private getNameView = (): JSX.Element => (
    <div className={classNames("participant-entry__name")}>
      <strong>{this.props.participant.name}</strong>
    </div>
  );

  private getAvatarImage = (): JSX.Element => (
    <img
      src={this.props.participant.avatar.url}
      alt={this.props.participant.avatar.altText}
    />
  );

  private getAvatarView = (): JSX.Element => (
    <div className={classNames("participant-entry__avatar")}>
      {this.getAvatarImage()}
    </div>
  );

  public render = (): JSX.Element => {
    return (
      <div className={classNames("participant-entry")}>
        {this.getAvatarView()}
        {this.getNameView()}
      </div>
    );
  };
}

export default ParticipantEntry;
