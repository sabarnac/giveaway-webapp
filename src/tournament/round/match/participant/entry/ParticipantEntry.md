Usage:

```tsx
import Participant from "./store/round/match/participant/Participant";

const participant: Participant = new Participant("FOO");
<ParticipantEntry participant={participant} inverse={false} />;
```
