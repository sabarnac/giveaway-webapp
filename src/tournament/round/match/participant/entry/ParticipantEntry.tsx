import React from "react";
import { observer } from "mobx-react";
import "./ParticipantEntry.scss";
import classNames from "classnames";
import Participant from "../../../../../store/round/match/participant/Participant";

/**
 * Properties of the participant entry React component.
 */
interface ParticipantEntryProps {
  /** The participant details. */
  participant: Participant;
  /** Whether to invert the color of the participant view or not. */
  invert?: boolean;
}

/**
 * React component for the participant entry.
 */
export default observer(
  (props: ParticipantEntryProps): JSX.Element => (
    <div
      className={classNames("participant", "participant-entry", {
        "participant--invert": props.invert,
        "participant-entry--invert": props.invert
      })}
    >
      <div className={classNames("participant-entry__avatar")}>
        <img
          src={props.participant.avatar.url}
          alt={props.participant.avatar.altText}
        />
      </div>
      <div className={classNames("participant-entry__name")}>
        <strong>{props.participant.properName}</strong>
      </div>
    </div>
  )
);
