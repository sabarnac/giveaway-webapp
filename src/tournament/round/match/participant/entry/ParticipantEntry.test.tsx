import React from "react";
import ParticipantEntry from "./ParticipantEntry";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyParticipant } from "../../../../../util/test/index";
import Participant from "../../../../../store/round/match/participant/Participant";

describe("ParticipantEntry Component.", () => {
  it("Matches snapshot.", () => {
    const participant: Participant = createDummyParticipant();
    const component: ReactTestRenderer = create(
      <ParticipantEntry participant={participant} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
