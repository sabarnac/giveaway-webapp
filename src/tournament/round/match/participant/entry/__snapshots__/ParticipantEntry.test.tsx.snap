// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticipantEntry Component. Matches snapshot. 1`] = `
<div
  className="participant participant-entry"
>
  <div
    className="participant-entry__avatar"
  >
    <img
      alt="fozbaz-1"
      src="https://api.adorable.io/avatars/480/fozbaz-1"
    />
  </div>
  <div
    className="participant-entry__name"
  >
    <strong>
      Fozbaz 1
    </strong>
  </div>
</div>
`;
